### PR TITLE
Vecadd demo

### PR DESCRIPTION
**Pull request by jpsamaroo**
This PR pulls in https://github.com/jpsamaroo/HSARuntime.jl for access to runtime calls and the all-important `HSAArray` type. A simple test is in place and working up until the attempt to compile and call the kernel. Once that's fixed we should be good to merge!

### CODE REVIEW

**Comment by jpsamaroo** [comment]
Status update:

This is going pretty well; I'm now at the point (locally) where I can get all the way through to being ready to actually launch the kernel. However, because of issues with generating a valid kernel when using `HSAArray`s as inputs, I've switched to a dummy kernel that takes only `Int32`s as arguments. I'll obviously change this once I've gotten launching squared away.

Currently, I need to do a few things before this is in any way ready to merge:
* Extract the kernel's symbol from the object code - I'm hoping that HSA will be cooperative and just go with what I hand it; if not, this might get quite annoying!
* Pass kernel arguments with the correct memory layout - Hopefully a `Tuple` of arguments will work here, at least for `Int32`s for now. `HSAArray` arguments may require some extra work.
* Switch back to `HSAArray`s - Aside from the above, I need to figure out the right way to do indexing into these arrays both locally and on the device. With them being shared virtual memory, the hope is that any indexing scheme that works on the host should work on the device equally well.

**Comment by vchuravy** [comment]
> With them being shared virtual memory, the hope is that any indexing scheme that works on the host should work on the device equally well.

Otherwise you might be able to go down the route that CUDAnative uses, with a translation to GPU-compatible arguments. That tends to be rather helpful to simplify codegen.

**Comment by jpsamaroo** [comment]
Do you have a link to where I should look at specifically in the CUDAnative codebase to see where this is done? I noticed a single Adapt rule for `Base.RefValue`, but nothing specific to device arrays.

**Comment by maleadt** [comment]
> Do you have a link to where I should look at specifically in the CUDAnative codebase to see where this is done? I noticed a single Adapt rule for `Base.RefValue`, but nothing specific to device arrays.

The adaption is registered in CuArrays.jl: https://github.com/JuliaGPU/CuArrays.jl/blob/f7a77634dfd9a16aeeae9ac7ff14a2cd24a9b58b/src/array.jl#L165-L166

`cudaconvert` (which uses Adapt) is called in CUDAnative: https://github.com/JuliaGPU/CUDAnative.jl/blob/23e10696bbab4501d578c67f2fe1089538adf3da/src/execution.jl#L168

Finally, CUDAdrv does `cconvert`/`unsafe_convert` as is expected from `ccall`-like APIs (but that isn't currently used for GPU conversions): https://github.com/JuliaGPU/CUDAdrv.jl/blob/5414ae8322ea6e66f20128e48c58cab1d11ee069/src/execution.jl#L161-L170

**Comment by jpsamaroo** [comment]
Thanks Tim! At this point I'm stuck on a GPU segfault when trying to access `c[1]`, which should be totally valid. Weirdly, while the DevicePointer I'm getting for `c` is always `0x1013000`, on both the host and device (as it's supposed to be, since host-allocated pointers should be able to be dereferenced on the device just fine), the GPU segfault indicates it's occurring somewhere in the range of `0x404C0xxx`. I can move that number around by changing the index into `c` that I try to read/write.

If either of you has any ideas for why this might be occurring (maybe something to do with the addrspaces in play), I would appreciate suggestions of where to focus my debug efforts.

**Comment by jpsamaroo** [comment]
PR is passing tests locally :confetti_ball: We'll probably need to fixup CI afterwards, since it seems like it might be some peculiarity with Ubuntu's ROCm apt packaging.

Thanks to @vchuravy and @maleadt for all the much-needed help :smile_cat:

File: src/compiler/driver.jl [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    end
+
+    prepare_execution!(ctx, mod)
+
+    check_invocation(ctx, entry)
+
+    # check generated IR
+    check_ir(ctx, mod)
+    verify(mod)
+    @show entry
```
Comment (vchuravy) [inline src/compiler/driver.jl]: left over show

File: src/compiler/irgen.jl [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+    triple!(mod, "amdgcn-amd-amdhsa")
+
+    # add debug info metadata
+    push!(metadata(mod), "llvm.module.flags",
+         MDNode([ConstantInt(Int32(1), JuliaContext()),    # llvm::Module::Error
+                 MDString("Debug Info Version"),
+                 ConstantInt(DEBUG_METADATA_VERSION(), JuliaContext())]))
+end
+
+# make function names safe for PTX
```
Comment (vchuravy) [inline src/compiler/irgen.jl]: Does GCN have the same restrictions?

File: src/compiler/irgen.jl [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+    LLVM.name!(entry, llvmfn)
+
+    # minimal required optimization
+    ModulePassManager() do pm
+        global global_ctx
+        global_ctx = ctx
+
+        add!(pm, ModulePass("LowerThrow", lower_throw!))
+        # TODO: Determine if these are useful, excise if not
+        #add!(pm, FunctionPass("HideUnreachable", hide_unreachable!))
```
Comment (vchuravy) [inline src/compiler/irgen.jl]: IIRC these work around limitations in ptxas

File: src/compiler/optim.jl [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        dispose(builder)
+    end
+
+    # early-inline the original entry function into the wrapper
+    push!(function_attributes(entry_f), EnumAttribute("alwaysinline", 0, JuliaContext()))
+    linkage!(entry_f, LLVM.API.LLVMInternalLinkage)
+
+    # set calling convention
+    #callconv!(entry_f, LLVM.API.LLVMCallConv(91))
+    callconv!(wrapper_f, LLVM.API.LLVMCallConv(91))
```
Comment (vchuravy) [inline src/compiler/optim.jl]: Document which one this is. Maybe we can have the enum in LLVM.jl

File: src/compiler/irgen.jl (hunk 2 above)
Comment (jpsamaroo) [reply src/compiler/irgen.jl]: Not sure, I'll have to go check the LLVM docs.

Comment (maleadt) [reply src/compiler/irgen.jl]: Those limitations aren't documented. Try throwing some strange function names at it and see if it fails :slightly_smiling_face:

Comment (jpsamaroo) [reply src/compiler/irgen.jl]: It seems to be okay with characters that aren't even in my terminal font. `vadd_σɷ̨ℵ⌣` works as expected.

File: src/compiler/irgen.jl (hunk 3 above)
Comment (jpsamaroo) [reply src/compiler/irgen.jl]: They also make my `dotrap` debug function not work, so no objections from me.

**Comment by vchuravy** [comment]
Further refactoring fixing CI can be done in smaller PRs!

**Pull request merged**